File: instrumentation/network/src/main/java/io/opentelemetry/android/instrumentation/network/NetworkApplicationListener.java
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.android.instrumentation.network;

import static io.opentelemetry.api.common.AttributeKey.stringKey;

import io.opentelemetry.android.instrumentation.common.ApplicationStateListener;
import io.opentelemetry.android.internal.services.network.CurrentNetworkProvider;
import io.opentelemetry.android.internal.services.network.NetworkChangeListener;
import io.opentelemetry.android.internal.services.network.data.CurrentNetwork;
import io.opentelemetry.api.common.AttributeKey;
import io.opentelemetry.context.Context;
import io.opentelemetry.instrumentation.api.instrumenter.Instrumenter;
import java.util.concurrent.atomic.AtomicBoolean;

class NetworkApplicationListener implements ApplicationStateListener {
    static final AttributeKey<String> NETWORK_STATUS_KEY = stringKey("network.status");

    private final CurrentNetworkProvider currentNetworkProvider;
    private final AtomicBoolean shouldEmitChangeEvents = new AtomicBoolean(true);

    NetworkApplicationListener(CurrentNetworkProvider currentNetworkProvider) {
        this.currentNetworkProvider = currentNetworkProvider;
    }

    void startMonitoring(Instrumenter<CurrentNetwork, Void> instrumenter) {
        currentNetworkProvider.addNetworkChangeListener(
                new TracingNetworkChangeListener(instrumenter, shouldEmitChangeEvents));
    }

    @Override
    public void onApplicationForegrounded() {
        shouldEmitChangeEvents.set(true);
    }

    @Override
    public void onApplicationBackgrounded() {
        shouldEmitChangeEvents.set(false);
    }

    private static final class TracingNetworkChangeListener implements NetworkChangeListener {

        private final Instrumenter<CurrentNetwork, Void> instrumenter;
        private final AtomicBoolean shouldEmitChangeEvents;

        TracingNetworkChangeListener(
                Instrumenter<CurrentNetwork, Void> instrumenter,
                AtomicBoolean shouldEmitChangeEvents) {
            this.instrumenter = instrumenter;
            this.shouldEmitChangeEvents = shouldEmitChangeEvents;
        }

        @Override
        public void onNetworkChange(CurrentNetwork currentNetwork) {
            if (!shouldEmitChangeEvents.get()) {
                return;
            }
            Context context = instrumenter.start(Context.current(), currentNetwork);
            instrumenter.end(context, currentNetwork, null, null);
        }
    }
}

File: instrumentation/network/src/main/java/io/opentelemetry/android/instrumentation/network/NetworkApplicationListener.kt
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.android.instrumentation.network

import io.opentelemetry.android.annotations.Incubating
import io.opentelemetry.android.common.internal.features.networkattributes.data.CurrentNetwork
import io.opentelemetry.android.internal.services.applifecycle.ApplicationStateListener
import io.opentelemetry.android.internal.services.network.CurrentNetworkProvider
import io.opentelemetry.android.internal.services.network.NetworkChangeListener
import io.opentelemetry.android.ktx.setSessionIdentifiersWith
import io.opentelemetry.android.session.SessionProvider
import io.opentelemetry.api.common.AttributeKey
import io.opentelemetry.api.common.Attributes
import io.opentelemetry.api.logs.Logger
import java.util.concurrent.atomic.AtomicBoolean
import java.util.function.Consumer

val NETWORK_STATUS_KEY: AttributeKey<String> = AttributeKey.stringKey("network.status")

@OptIn(Incubating::class)
internal class NetworkApplicationListener(
    private val currentNetworkProvider: CurrentNetworkProvider,
    private val sessionProvider: SessionProvider,
) : ApplicationStateListener {
    private val shouldEmitChangeEvents = AtomicBoolean(true)

    fun startMonitoring(
        eventLogger: Logger,
        additionalExtractors: List<NetworkAttributesExtractor>,
    ) {
        currentNetworkProvider.addNetworkChangeListener(
            TracingNetworkChangeListener(
                eventLogger,
                sessionProvider,
                shouldEmitChangeEvents,
                additionalExtractors,
            ),
        )
    }

    override fun onApplicationForegrounded() {
        shouldEmitChangeEvents.set(true)
    }

    override fun onApplicationBackgrounded() {
        shouldEmitChangeEvents.set(false)
    }

    private class TracingNetworkChangeListener(
        private val eventLogger: Logger,
        private val sessionProvider: SessionProvider,
        private val shouldEmitChangeEvents: AtomicBoolean,
        private val additionalExtractors: List<NetworkAttributesExtractor>,
    ) : NetworkChangeListener {
        override fun onNetworkChange(currentNetwork: CurrentNetwork) {
            if (!shouldEmitChangeEvents.get()) {
                return
            }
            val attributesBuilder = Attributes.builder()
            additionalExtractors.forEach(
                Consumer { extractor: NetworkAttributesExtractor ->
                    extractor(attributesBuilder, currentNetwork)
                },
            )
            val builder = eventLogger.logRecordBuilder()
            builder
                .setSessionIdentifiersWith(sessionProvider)
                .setEventName("network.change")
                .setAllAttributes(attributesBuilder.build())
                .emit()
        }
    }
}
